chips/nrf52/src/i2c.rs

```diff
@@ -0,0 +1,446 @@
+                Some(client) => {
+                    match self.buf.take() {
+                        None => (),
+                        Some(buf) => {
+                            client.command_complete(buf, errorsrc.into());
+                        }
+                    }
+                }
+            };
+        }
+
+        if self.regs().events_suspended.get() == 1 {
+            self.regs().events_suspended.set(0);
+        }
+
+        if self.regs().events_rxstarted.get() == 1 {
+            self.regs().events_rxstarted.set(0);
+        }
+
+        if self.regs().events_lastrx.get() == 1 {
+            self.regs().events_lastrx.set(0);
+        }
+
+        if self.regs().events_lasttx.get() == 1 {
+            self.regs().events_lasttx.set(0);
+        }
+
+        self.busy.set(false);
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.enabled.get()
+    }
+}
+
+impl hil::i2c::I2CMaster for TWIM {
+    fn enable(&self) {
+        self.enable();
+    }
+
+    fn disable(&self) {
+        self.disable();
+    }
+
+    fn write_read(&self, addr: u8, data: &'static mut [u8], write_len: u8, read_len: u8) {
+        self.regs().address.set((addr >> 1) as u32);
+        self.regs().txd_ptr.set(data.as_mut_ptr());
+        self.regs().txd_maxcnt.set(write_len as u32);
+        self.regs().rxd_ptr.set(data.as_mut_ptr());
+        self.regs().rxd_maxcnt.set(read_len as u32);
+        self.regs().shorts.set({
+            let mut shorts = registers::Shorts(0);
+            shorts.set_lasttx_startrx(1);
+            shorts.set_lastrx_stop(1);
+            shorts
+        });
+        self.regs().intenset.set({
+            let mut intenset = registers::InterruptEnable(0);
+            intenset.set_stopped(1);
+            intenset.set_error(1);
+            intenset
+        });
+        // start the transfer
+        self.busy.set(true);
+        self.regs().tasks_starttx.set(1);
+        self.buf.replace(data);
+    }
+
+    fn write(&self, addr: u8, data: &'static mut [u8], len: u8) {
+        self.regs().address.set((addr >> 1) as u32);
+        self.regs().txd_ptr.set(data.as_mut_ptr());
+        self.regs().txd_maxcnt.set(len as u32);
+        self.regs().shorts.set({
+            let mut shorts = registers::Shorts(0);
+            shorts.set_lasttx_stop(1);
+            shorts
+        });
+        self.regs().intenset.set({
+            let mut intenset = registers::InterruptEnable(0);
+            intenset.set_stopped(1);
+            intenset.set_error(1);
+            intenset
+        });
+        // start the transfer
+        self.busy.set(true);
+        self.regs().tasks_starttx.set(1);
+        self.buf.replace(data);
+    }
+
+    fn read(&self, addr: u8, buffer: &'static mut [u8], len: u8) {
+        self.regs().address.set((addr >> 1) as u32);
+        self.regs().rxd_ptr.set(buffer.as_mut_ptr());
+        self.regs().rxd_maxcnt.set(len as u32);
+        self.regs().shorts.set({
```

Comment:
Since shorts are a bit specific to the NRF52, can you please document how you're using them functionally? i.e, an inline comment saying something like "automatically trigger the stop event when done reading"

Reply:
I pushed a comment about the shortcuts used.